gusto/limiters.py

```diff
@@ -0,0 +1,117 @@
+from __future__ import absolute_import, print_function, division
+from firedrake import dx
+from firedrake.function import Function
+from firedrake.functionspace import FunctionSpace
+from firedrake.parloops import par_loop, READ, RW, INC
+from firedrake.slope_limiter.vertex_based_limiter import VertexBasedLimiter
+
+__all__ = ["ThetaLimiter"]
+
+_copy_into_Q1DG_loop = """
+theta_hat[0][0] = theta[0][0];
+theta_hat[1][0] = theta[1][0];
+theta_hat[2][0] = theta[3][0];
+theta_hat[3][0] = theta[4][0];
+"""
+_copy_from_Q1DG_loop = """
+theta[0][0] = theta_hat[0][0];
+theta[1][0] = theta_hat[1][0];
+theta[3][0] = theta_hat[2][0];
+theta[4][0] = theta_hat[3][0];
+"""
+
+_check_midpoint_values_loop = """
+if (theta[2][0] > fmax(theta[0][0], theta[1][0]))
+    theta[2][0] = 0.5 * (theta[0][0] + theta[1][0]);
+else if (theta[2][0] < fmin(theta[0][0], theta[1][0]))
+    theta[2][0] = 0.5 * (theta[0][0] + theta[1][0]);
+if (theta[5][0] > fmax(theta[3][0], theta[4][0]))
+    theta[5][0] = 0.5 * (theta[3][0] + theta[4][0]);
+else if (theta[5][0] < fmin(theta[3][0], theta[4][0]))
+    theta[5][0] = 0.5 * (theta[3][0] + theta[4][0]);
+"""
+
+_weight_kernel = """
+for (int i=0; i<weight.dofs; ++i) {
+    weight[i][0] += 1.0;
+    }"""
+
+_average_kernel = """
+for (int i=0; i<vrec.dofs; ++i) {
+        vrec[i][0] += v_b[i][0]/weight[i][0];
+        }"""
+
+
+class ThetaLimiter(object):
```

Comment:
You might want to have a look at `limiter.py` and `vertex_based_limiter.py` in `firedrake/firedrake/slope_limiters`. With some modifications, your `ThetaLimiter` could inherit from either the `Limiter` class (or possibly `VertexBasedLimiter`). That way, your limiter conforms to the existing limiter setup in current Firedrake.

Reply:
It used to do this in previous commits, but I ended up thinking that it wouldn't be appropriate. The Firedrake `Limiter` class has several `@abstractmethod`s that we don't use in this limiter.